subsys/power/device_pm.c

```diff
@@ -0,0 +1,301 @@
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->async_request = target_state;
+		dev->pm->async_req_status = DEVPM_ASYNC_REQ_QUEUED;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+
+		return DEVICE_PM_RET_QUEUED;
+	}
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
+	ret = set_dev_pm_state(dev, target_state);
+	if (ret < 0) {
+		LOG_ERR("Set power state(%d) error(%d)\n",
+						target_state, ret);
+	}
+	k_sem_give(&dev->pm->pm_lock);
+
+handler_out:
+	return ret;
+}
+
+static int device_pm_handle_get(struct device *dev, u32_t pm_flags)
+{
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		ret = -ERANGE;
+		goto get_out;
+	}
+	dev->pm->usage++;
+
+	if (!dev->pm->enable) {
+		goto get_out;
+	}
+
+	if (dev->pm->usage > 0) {
+		ret = device_pm_state_handler(dev,
+				DEVICE_PM_ACTIVE_STATE, pm_flags);
+	}
+
+
+get_out:
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+static int device_pm_handle_put(struct device *dev, u32_t pm_flags)
+{
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage <= 0) {
+		LOG_ERR("usage count underflow error!!\n");
+		ret = -ERANGE;
+		goto put_out;
+	}
+	dev->pm->usage--;
+
+	if (!dev->pm->enable) {
+		goto put_out;
+	}
+
+	if (dev->pm->usage < 1) {
+		ret = device_pm_state_handler(dev,
+				DEVICE_PM_SUSPEND_STATE, pm_flags);
+	}
+
+put_out:
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	return device_pm_handle_get(dev, DEVICE_PM_ASYNC);
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	return device_pm_handle_get(dev, 0);
+}
+
+int device_pm_put(struct device *dev)
+{
+	return device_pm_handle_put(dev, DEVICE_PM_ASYNC);
+}
+
+int device_pm_put_sync(struct device *dev)
+{
+	return device_pm_handle_put(dev, 0);
+}
+
+void device_pm_enable(struct device *dev)
+{
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->enable = true;
+
+	/* During the driver init, device can set the
+	 * PM state accordingly. For later cases we need
+	 * to check the usage and set the device PM state.
+	 */
+	if (!dev->pm->dwork.work.handler) {
+		dev->pm->dev = dev;
+		k_delayed_work_init(&dev->pm->dwork, pm_work_handler);
+	} else {
+		if (dev->pm->usage) {
+			ret = device_pm_state_handler(dev,
+				DEVICE_PM_ACTIVE_STATE, DEVICE_PM_ASYNC);
+		} else {
+			ret = device_pm_state_handler(dev,
+				DEVICE_PM_SUSPEND_STATE, DEVICE_PM_ASYNC);
+		}
+		if (ret < 0) {
+			LOG_ERR("Device PM state set error!!\n");
+		}
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+void device_pm_disable(struct device *dev)
+{
```

Comment:
IMHO here we should wake-up the device regardless of the usage count.

Reply:
Ok got it.